nomad/client_csi_endpoint.go

```diff
@@ -109,6 +109,96 @@ func (a *ClientCSI) ControllerDetachVolume(args *cstructs.ClientCSIControllerDet
 	return fmt.Errorf("controller detach volume: %v", err)
 }
 
+func (a *ClientCSI) ControllerCreateVolume(args *cstructs.ClientCSIControllerCreateVolumeRequest, reply *cstructs.ClientCSIControllerCreateVolumeResponse) error {
+	defer metrics.MeasureSince([]string{"nomad", "client_csi_controller", "create_volume"}, time.Now())
+
+	clientIDs, err := a.clientIDsForController(args.PluginID)
+	if err != nil {
+		return fmt.Errorf("create volume: %v", err)
+	}
+
+	for _, clientID := range clientIDs {
+		args.ControllerNodeID = clientID
+		state, ok := a.srv.getNodeConn(clientID)
+		if !ok {
+			return findNodeConnAndForward(a.srv,
+				clientID, "ClientCSI.ControllerCreateVolume", args, reply)
+		}
+
+		err = NodeRpc(state.Session, "CSI.ControllerCreateVolume", args, reply)
+		if err == nil {
+			return nil
+		}
+		if a.isRetryable(err) {
+			a.logger.Debug("failed to reach controller on client",
+				"nodeID", clientID, "err", err)
+			continue
+		}
+		return fmt.Errorf("create volume: %v", err)
+	}
+	return fmt.Errorf("create volume: %v", err)
+}
```

Comment:
I tried that at first but ran into trouble trying to pass in the `args` because they're all different structs. Probably something clever I could do with a closure there. Lemme take another crack at it.

Reply:
Can you args just be `interface{}` - that's what `NodeRpc` and `findNodeConnAndForward` expect?

Reply:
We have to modify the arg per client... I came up with this though and I think it reads nicely. Will cut out a bunch of repeated code once I call it across the whole file.

```go

func (a *ClientCSI) ControllerAttachVolume(args *cstructs.ClientCSIControllerAttachVolumeRequest, reply *cstructs.ClientCSIControllerAttachVolumeResponse) error {
	defer metrics.MeasureSince([]string{"nomad", "client_csi_controller", "attach_volume"}, time.Now())

	err := a.sendCSINodeRPC(args.PluginID,
		"ClientCSI.ControllerAttachVolume", args, reply,
		func(clientID string) {
			args.ControllerNodeID = clientID
		})
	if err != nil {
		return fmt.Errorf("controller attach volume: %v", err)
	}
	return nil
}

func (a *ClientCSI) sendCSINodeRPC(pluginID string, method string, args interface{}, reply interface{}, applyFn func(clientID string)) error {

	clientIDs, err := a.clientIDsForController(pluginID)
	if err != nil {
		return err
	}

	for _, clientID := range clientIDs {
		applyFn(clientID)
		state, ok := a.srv.getNodeConn(clientID)
		if !ok {
			return findNodeConnAndForward(a.srv,
				clientID, method, args, reply)
		}

		err = NodeRpc(state.Session, method, args, reply)
		if err == nil {
			return nil
		}
		if a.isRetryable(err) {
			a.logger.Debug("failed to reach controller on client",
				"nodeID", clientID, "err", err)
			continue
		}
		return err
	}
	return err
}
```

Reply:
Oh I see.  I missed that args modification.  I guess we can also introduce an `ClientCSIRequest` interface with `SetClientID` method, and eliminate that applyfn. Up to you.

Reply:
That's a nice update. Will do.

Reply:
Done in 4125f6a